modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -40,108 +41,180 @@
 import org.elasticsearch.index.mapper.internal.VersionFieldMapper;
 import org.elasticsearch.script.CompiledScript;
 import org.elasticsearch.script.ExecutableScript;
+import org.elasticsearch.script.Script;
 import org.elasticsearch.script.ScriptContext;
 import org.elasticsearch.script.ScriptService;
 import org.elasticsearch.search.SearchHit;
 import org.elasticsearch.search.SearchHitField;
 import org.elasticsearch.threadpool.ThreadPool;
 
 import java.util.HashMap;
-import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.function.BiFunction;
 
 import static java.util.Collections.emptyMap;
 
 /**
  * Abstract base for scrolling across a search and executing bulk indexes on all
  * results.
  */
-public abstract class AbstractAsyncBulkIndexByScrollAction<Request extends AbstractBulkIndexByScrollRequest<Request>>
-        extends AbstractAsyncBulkByScrollAction<Request, BulkIndexByScrollResponse> {
+public abstract class AbstractAsyncBulkIndexByScrollAction<Request extends AbstractBulkByScrollRequest<Request>>
+        extends AbstractAsyncBulkByScrollAction<Request> {
 
-    private final ScriptService scriptService;
-    private final CompiledScript script;
+    protected final ScriptService scriptService;
+    protected final ClusterState clusterState;
+    private final ScriptApplier scriptApplier;
```

Comment:
If you made this a BiFunction then you could get away with the noop implementation being just `(a, b) -> a` and save a ton of code! `buildScriptApplier` would have to return a BiFunction as well, but extensions to this class could still build an extension of ScriptApplier just like they do now.